contributors/design-proposals/auth/svcacct-token-volume-source.md

```diff
@@ -0,0 +1,138 @@
+# Service Account Token Volumes
+
+Authors:
+  @smarterclayton
+  @liggitt
+  @mikedanese
+
+## Summary
+
+Kubernetes is able to provide pods with unique identity tokens that can prove
+the caller is a particular pod to a Kubernetes API server. These tokens are
+injected into pods as secrets. This proposal proposes a new mechanism of
+distribution with support for [improved service account
+tokens](https://github.com/kubernetes/community/pull/1460) and explores how to
+migrate from the existing mechanism backwards compatibly.
+
+## Motivation
+
+Many workloads running on Kubernetes need to prove to external parties who they
+are in order to participate in a larger application environment. This identity
+must be attested to by the orchestration system in a way that allows a third
+party to trust that an arbitrary container on the cluster is who it says it is.
+In addition, infrastructure running on top of Kubernetes needs a simple
+mechanism to communicate with the Kubernetes APIs and to provide more complex
+tooling. Finally, a significant set of security challenges are associated with
+storing service account tokens as secrets in Kubernetes and limiting the methods
+whereby malicious parties can get access to these tokens will reduce the risk of
+platform compromise.
+
+As a platform, Kubernetes should evolve to allow identity management systems to
+provide more powerful workload identity without breaking existing use cases, and
+provide a simple out of the box workload identity that is sufficient to cover
+the requirements of bootstrapping low-level infrastructure running on
+Kubernetes. We expect that other systems to cover the more advanced scenarios,
+and see this effort as necessary glue to allow more powerful systems to succeed.
+
+With this feature, we hope to provide a backwards compatible replacement for
+service account tokens that strengthens the security and improves the
+scalability of the platform.
+
+## Proposal
+
+Kubernetes should:
+
+1.  Allow the automatic service account token injection to be disabled via
+    configuration.
+1.  Allow automatic service account token creation to occur only when a user
+    explicitly requests.
+1.  Implement a ServiceAccountToken volume projection that that maintains a
+    service account token requested by the node from the TokenRequest API.
+
+### Token Volume Projection
+
+A new volume projection will be implemented with an API that closely matches the
+TokenRequest API.
+
+```go
+type ServiceAccountTokenProjection struct {
```

Comment:
I don't see the name of the service account here.  Have you considered making it so that a Pod can use multiple SAs at the same time?  I can see scenarios where, say, a set of microservices get merged into one service and the things these things are talking to other services with unique SAs. Having the new merged service be able to use the existing SAs would be nice.

Reply:
Nothing prevents you mounting one of their secrets today, so i think it's reasonable to extend that.  I think it's a reasonable addition, but is also something that could be added later without impact (an optional field where the default is the pod's service account).  Do you think a ton of people have this right now Joe?  The only example I've seen has been prometheus which has a scraper token and a service account token - I would definitely use something like that there, but that's 1 out of several hundred apps I've seen people deploy, so i'd suspect its 5% case (a useful one).  I don't *think* any of the other core services have come up with this use case yet.

Reply:
I also think that this may be more interesting as people start mapping to Cloud SAs.  You may have one SA for talking within the cluster and another SA to talk to the cloud.  Ideally these would be the same but it depends on how flexible the mappings are.

I'm just allergic to any 1:1 bindings for stuff like this. In my experience it almost always comes back to bite you.  And if we don't have it in there earlier than folks will tend to work with what they have vs. come back and add something.  However, I wouldn't lay down on the tracks if this didn't make it in the first version.

I think it is also worth calling out (perhaps in an example) that a pod can project the same SA multiple times with different audiences. Without making this clear people will naively just add more audiences and pass that token around. This kind of defeats the whole purpose of the audience field. In my mind having multiple audiences on a token is a bad smell and we shouldn't make that the easy/suggested/obvious case.

Reply:
This would add a lot of questions to the design (e.g. how do these extra service accounts interact with PodSecurityPolicy, ImagePullSecrets attached to the service account, etc...). As Clayton says, this functionality could be added backwards compatibly later and I would prefer to solve this with a holistic design for sub-pod isolation.

Reply:
> think it is also worth calling out (perhaps in an example) that a pod can project the same SA multiple times with different audiences. Without making this clear people will naively just add more audiences and pass that token around.

Good point.